fix: events index: record processed epochs and tipsets for events and eth_get_log blocks till requested tipset has been indexed

aarshkshah1992 opened the pull request:
Closes https://github.com/filecoin-project/lotus/issues/11640 and #12078

Stebalien commented on chain/events/filter/index.go:
```diff
@@ -75,20 +79,22 @@ var ddls = []string{
 	insertEntry          = `INSERT OR IGNORE INTO event_entry(event_id, indexed, flags, key, codec, value) VALUES(?, ?, ?, ?, ?, ?)`
 	revertEventsInTipset = `UPDATE event SET reverted=true WHERE height=? AND tipset_key=?`
 	restoreEvent         = `UPDATE event SET reverted=false WHERE height=? AND tipset_key=? AND tipset_key_cid=? AND emitter_addr=? AND event_index=? AND message_cid=? AND message_index=?`
+	insertEventsSeen     = `INSERT OR IGNORE INTO events_seen(height, tipset_key, tipset_key_cid) VALUES(?, ?, ?)`
```
Any reason we're storing both the tipset key and cid? I guess this applies to the events table as well.

rvagg replied:
This is a good question for the rest of the schema actually, and @aarshkshah1992 is just conforming to what's there - we keep both in `event`, we insert them both with new events and we query matching them both for "exists" and when we do a "restore" operation. Maybe we should just ditch tipset_key and only ever use tipset_key_cid? I'm not seeing what value-add both are giving us in here.

rvagg replied:
@aarshkshah1992 can you look at the possibility of removing the key and just using the CID for this particular case, we can deal with the other instances later I think.

aarshkshah1992 replied:
Changed this to just use the CID. Whenever we have they key, we also have the CID ofcourse.

aarshkshah1992 replied:
This is done.

Stebalien commented on chain/events/filter/index.go:
```diff
@@ -520,6 +609,13 @@ func (ei *EventIndex) Close() error {
 }
 
+func (ei *EventIndex) isTipsetProcessed(ctx context.Context, tipsetKey []byte) (bool, error) {
+	row := ei.db.QueryRowContext(ctx, "SELECT COUNT(*) > 0 FROM events_seen WHERE tipset_key = ?", tipsetKey)
```
I think we need an additional "reverted" flag if we're going to use this table to fix https://github.com/filecoin-project/lotus/issues/11640. The issue here is:

1. Apply TipSet A at epoch X (is process returns false).
2. Switch to TipSet B at epoch X (is processed returns false).
4. Switch back to TipSetA (is processed returns _true_, but it's not).

When I discussed this with @rvagg the plan was (IIRC) to record:

1. The earliest processed tipset/epoch for backfilling.
2. The last processed tipset (and _only_ the last processed one).

When "catching up", you'd always look at the "path" between the last processed tipset and the current tipset (taking reverts into account).

But that doesn't solve your `getLogs` case. If a user asks for a historical "uncle", there's no easy way to tell if that uncle has been processed (you don't know what chain it's on).

So, I think the solution is to add a reverted flag so, on start, we can find the latest non-reverted tipset and "catch up" from there.

---

Note: We don't have to do this now, but I'd like to clean all this up a bit:

1. Remove the reverted flag from the `events` table, that'll be recorded in the `events_seen` table (maybe call this `tipsets`?).
2. Remove the `tipset_cid`/`tipset_key` from the `events` table as well and reference the correct entry in the `tipsets` table by foreign key.

It makes some of the queries more complex, but the events table is getting fat and we're doing more work than necessary to process reverts.

rvagg replied:
I'm not sure we need a "reverted" column specifically for this use-case, we just need to know whether a particular tipset has been inserted or not. Even if it flips from applied to reverted and back to applied again, the tipset doesn't change so it's events won't change. What am I not seeing that #11640 would need a "reverted" @Stebalien? It makes sense if we have a "tipset" table where we store all the tipset things, but that's probably too much scope for this.

Stebalien replied:
You're right, it fixes the main issue but not my followup comment (https://github.com/filecoin-project/lotus/issues/11640#issuecomment-1964955238).

1. Main issue: Sometimes a user will call `getLogs` on a tipset we haven't processed yet. This PR fixes that case (or works towards fixing it).
2. Followup comment: Sometimes we'll shut down before indexing some tipset and, when we restart, we need to "catch up" and index any missing tipsets. In this case, my concern is that we'll shutdown right after step 3 above ("Switch back to TipSetA"), restart, and think "oh, I see we've already indexed TipSetA". But we _actually_ need to go and un-mark TipSetA as reverted and mark TipSetB as reverted instead.

aarshkshah1992 replied:
I think I agree with Steb.

Being able to answer "What's the latest tipset for which you have events on the cannonical chain" is important. Might even be for this work itself. Still working through it. But have made this change.

aarshkshah1992 replied:
This is done.

Stebalien reviewed (commented):
I have some opinions but this really needs a look by @rvagg (who has done more thinking on this than I have at this point).

rvagg commented on chain/events/filter/index.go:
```diff
@@ -407,6 +429,63 @@ func (ei *EventIndex) migrateToVersion5(ctx context.Context) error {
+    FROM event e
+    WHERE NOT EXISTS (
+        SELECT 1 FROM events_seen es
+        WHERE es.height = e.height AND es.tipset_key = e.tipset_key AND es.tipset_key_cid = e.tipset_key_cid
```
an alternative would be to find the min and max and fill in all between those two as being seen, that might yield a a state that's much more likely to be valid; there's no good reason to have blanks between min and max, but there are good reasons to have missing events, (this can been seen particularly on calibnet)

aarshkshah1992 replied:
@rvagg min and max of what ? The problem is that for epochs that do not have events, we simply do not have any entry for them in any of our event tabls.

rvagg replied:
I was thinking min and max height; start from `MIN(height)` and add an entry for every one until `MAX(height)` .. but I suppose you need tipset_key_cid for each of those too so you'd have to go looking for it - but this could become a chain walk operation too -- find min and max, walk chain between and add an entry for each one.

Maybe this isn't worth thinking too hard about since it's a migration but I _think_ doing a walk and including _every_ epoch in that range is much more likely to be accurate than just filling in the ones we have events for. Because we should have an entry in the table even for tipsets that have no events because that's part of what we're solving here - no events in the tipset, vs no events in the database but maybe tipset has events?

rvagg commented on chain/events/filter/index.go:
```diff
@@ -407,6 +429,63 @@ func (ei *EventIndex) migrateToVersion5(ctx context.Context) error {
+		return xerrors.Errorf("commit transaction: %w", err)
+	}
+
+	ei.vacuumDBAndCheckpointWAL(ctx)
```
this needs to be removed from `migrateToVersion5`, we bump it to the last migration so if you're doing >1 migrations we don't do this more than once (and yes, we should do this better now we have so many migrations but it works for now)

aarshkshah1992 replied:
Fixed this.

rvagg commented on chain/events/filter/index.go:
```diff
@@ -520,6 +640,61 @@ func (ei *EventIndex) Close() error {
+}
+
+func (ei *EventIndex) IsTipsetProcessed(ctx context.Context, tipsetKeyCid []byte) (bool, error) {
+	row := ei.db.QueryRowContext(ctx, "SELECT COUNT(*) > 0 FROM events_seen WHERE tipset_key_cid = ?", tipsetKeyCid)
```
this should probably be a prepared statement

aarshkshah1992 replied:
Done.

rvagg commented on chain/events/filter/index.go:
```diff
@@ -520,6 +640,61 @@ func (ei *EventIndex) Close() error {
 	return ei.db.Close()
 }
 
+func (ei *EventIndex) SubscribeUpdates() (uint64, chan *EventIndexUpdate) {
```
perhaps you should just return an `Unsubscribe()` or `Cancel()` function from this instead of a uint64 so you don't need a second method for `UnsubscribeUpdates`, and you also don't allow for one caller to cancel the subscription of another

aarshkshah1992 replied:
Done.

rvagg commented on chain/events/filter/index.go:
```diff
@@ -530,16 +705,45 @@ func (ei *EventIndex) CollectEvents(ctx context.Context, te *TipSetEvents, rever
 		}
 
+		ei.mu.Lock()
+		tSubs := ei.updateSubs
```
this isn't a full copy, you'll need to go further than this to be protected

```suggestion
		tSubs := make([]int, len(ei.updateSubs))
		copy(tSubs, ei.updateSubs)
```

rvagg replied:
could make this a private method `updateSubsCopy()` or similar, since it's done twice in here

aarshkshah1992 replied:
Done.

rvagg commented on chain/events/filter/index.go:
```diff
@@ -108,6 +127,25 @@ type EventIndex struct {
+}
+
+type EventIndexUpdate struct {
+	Height    uint64
```
this should probably be `abi.ChainEpoch`, I see some casting going on that would be redundant here

aarshkshah1992 replied:
Done.

rvagg commented on chain/events/filter/index.go:
```diff
@@ -108,6 +127,25 @@ type EventIndex struct {
+
+type EventIndexUpdate struct {
+	Height    uint64
+	TipsetCid cid.Cid
```
```suggestion
	TipSetKeyCid cid.Cid
```

aarshkshah1992 replied:
Done.

rvagg commented on chain/events/filter/index.go:
```diff
@@ -550,6 +754,10 @@ func (ei *EventIndex) CollectEvents(ctx context.Context, te *TipSetEvents, rever
 	if err != nil {
 		return xerrors.Errorf("load executed messages: %w", err)
 	}
+	tsKeyCid, err := te.msgTs.Key().Cid()
+	if err != nil {
+		return xerrors.Errorf("tipset key cid: %w", err)
+	}
```
this is done in both branches, pull up to the top and do once?

rvagg commented on chain/events/filter/index.go:
```diff
@@ -567,11 +775,6 @@ func (ei *EventIndex) CollectEvents(ctx context.Context, te *TipSetEvents, rever
 				addressLookups[ev.Emitter] = addr
 			}
 
-			tsKeyCid, err := te.msgTs.Key().Cid()
```
good catch btw, for _each_ event, for _each_ element of an event, we're encoding the tipset key 🤦

aarshkshah1992 replied:
Done.

rvagg commented on chain/events/filter/index.go:
```diff
@@ -646,16 +849,61 @@ func (ei *EventIndex) CollectEvents(ctx context.Context, te *TipSetEvents, rever
 					log.Warnf("restored %d events but expected only one to exist", rowsAffected)
 				}
+
+				res, err = tx.Stmt(ei.stmtRestoreEventSeen).Exec(
```
this block seems misplaced - restoring for the tipset inside the internal loop? shouldn't this be with the `stmtInsertEventsSeen` below? you could even do the insert only if `rowsAffected == 0` after this query

aarshkshah1992 replied:
Done.

rvagg commented on node/impl/full/eth.go:
```diff
@@ -1270,6 +1307,35 @@ func (e *EthEventHandler) EthGetLogs(ctx context.Context, filterSpec *ethtypes.E
 	return ethFilterResultFromEvents(ctx, ces, e.SubManager.StateAPI)
 }
 
+func (e *EthEventHandler) waitForHeightProcessed(ctx context.Context, height abi.ChainEpoch) error {
```
do you think it's safe enough to just rely on the RPC call timeout to time this out rather than having an explicit one in here?

aarshkshah1992 replied:
Have added a 3 epohc timeout here just in case.

aarshkshah1992 replied:
Done.

rvagg commented on node/impl/full/eth.go:
```diff
@@ -1270,6 +1307,35 @@ func (e *EthEventHandler) EthGetLogs(ctx context.Context, filterSpec *ethtypes.E
+		return nil
+	}
+
+	if height > e.Chain.GetHeaviestTipSet().Height() {
+		return xerrors.New("height is in the future")
+	}
```
move to the top?

aarshkshah1992 replied:
Done.

rvagg commented on node/impl/full/eth.go:
```diff
@@ -1270,6 +1307,35 @@ func (e *EthEventHandler) EthGetLogs(ctx context.Context, filterSpec *ethtypes.E
 
+func (e *EthEventHandler) waitForHeightProcessed(ctx context.Context, height abi.ChainEpoch) error {
+	ei := e.EventFilterManager.EventIndex
+	b, err := ei.IsHeightProcessed(ctx, uint64(height))
```
There's a race in here between calling `IsHeightProcessed` and `SubscribeUpdates` where the height may be processed between the two calls. You might need a `SubscribeUpdatesIfHeightProcessed` that can do it atomically for you; then you're going to have to make sure you hold a lock internally between the query and getting the channel in place.

A slight variation on this would be to ditch the "is height processed" check entirely, always subscribe, but immediately send the last processed tipset as soon as you've added the subscription (which also needs to be done while holding the lock) so that you always get a `EventIndexUpdate` with the current head which may be what you want, so it's like "is height processed" but we find out while watching the channel, which should have an event waiting for us.

aarshkshah1992 replied:
Fixed this by making the event stateless and making all clients defer to the DB as the source of truth after being notified that the DB has been updated.

rvagg commented on node/impl/full/eth.go:
```diff
@@ -44,6 +44,11 @@ var ErrUnsupported = errors.New("unsupported method")
 
+var (
+	// wait for 3 epochs
+	eventReadTimeout = 90 * time.Second
```
there's a `EpochDurationSeconds` in chain/actors/builtin

aarshkshah1992 replied:
Fixed this.

rvagg replied:
did you forget to push the fix? still has `90` hardwired here

```suggestion
	eventReadTimeout = builtin.EpochDurationSeconds * time.Second
```

aarshkshah1992 replied:
@rvagg I did but the tests barf on this because the tests have like 2ms block times and it's sometimes not enough for the head to get processed.

rvagg replied:
hah, that's a bit of a dilemma then!

rvagg reviewed (approved):
I _think_ this looks ok, but it's slightly risky so I'd be more comfortable having someone else look too, @Stebalien if available would be good

aarshkshah1992 commented:
@rvagg Would love a review from @Stebalien too but which part do you find risky ?

Stebalien commented on node/impl/full/eth.go:
```diff
@@ -1258,10 +1263,47 @@ func (e *EthEventHandler) EthGetLogs(ctx context.Context, filterSpec *ethtypes.E
+
+	if pf.tipsetCid == cid.Undef {
+		maxHeight := pf.maxHeight
+		if maxHeight == -1 {
+			maxHeight = e.Chain.GetHeaviestTipSet().Height()
+		}
```
We should always check the heaviest tipset, IMO. If we don't have a tipset at or beyond the requested one, fail immediately instead of waiting.

aarshkshah1992 replied:
Added a check for this.

Stebalien commented on node/impl/full/eth.go:
```diff
@@ -1258,10 +1263,47 @@ func (e *EthEventHandler) EthGetLogs(ctx context.Context, filterSpec *ethtypes.E
+		if maxHeight == -1 {
+			maxHeight = e.Chain.GetHeaviestTipSet().Height()
+		}
+		err := e.waitForHeightProcessed(ctx, maxHeight)
```
I kind of assumed the event indexing system would handle this for us. Although the install/take/uninstall pattern makes that hard.

Maybe add a `GetEvents` method to the `EventFilterManager` that waits if necessary then returns the events (and doesn't require the install/uninstall dance)?

Stebalien replied:
Yeah, we should _really_ do it this way. You'll be able to drop all the subscribe stuff as well.

Basically:

1. Add a `sync.Cond` to `EventFilterManager`.
2. First up, check to make sure the latest tipset we have in the chainstore is beyond the requested tipset/height.
3. Take `EventFilterManager.mu`
4. Compare `currentHeight` against the target height.
5. If `currentHeight` is less than the target, wait on the condition and repeat step 4.
6. If greater than or equal to, check if we've actually processed the tipset we care about. If not, abort.
7. Release the lock and collect the relevant events.

Stebalien replied:
Hm. A cond variable is going to be hard if we want to obey the context.

But... I think there's a better way. We can:

1. Have event filters decide if they're done (as they process events).
2. Add a `WaitFinished` (or something) method to event filters that waits for them to declare that they're "done" (can match no more events.

aarshkshah1992 replied:
I would like to address this as part of https://github.com/filecoin-project/lotus/issues/12116 rather than solving this in a patchy way here. For now, this PR does solve the problem of `eth_getLogs` failing to return events for tipsets on the cannonical chain and I would like to ship this as an intermediate step since we're now anyways going to prioritise https://github.com/filecoin-project/lotus/issues/12116

Stebalien replied:
My proposal should be much simpler than the current subscribe logic, unless I'm missing something. This kind of subscribe pattern tends to have deadlock issues (although I haven't looked at your implementation closely beyond suggesting an alternative).

Stebalien replied:
Nvm. We discussed this and @aarshkshah1992's proposal is better overall.

Stebalien commented on node/impl/full/eth.go:
```diff
@@ -1258,10 +1263,47 @@ func (e *EthEventHandler) EthGetLogs(ctx context.Context, filterSpec *ethtypes.E
+		if !b {
+			return nil, xerrors.Errorf("event index failed to index tipset %s", pf.tipsetCid.String())
+		}
+	}
```
We need to check the range (min/max should be sufficient as long as we assume automatic backfilling), not just the end.

aarshkshah1992 replied:
Assuming autobackfilling works, the presence of events for  `max`  implies the presence of events for `min` no ?

Auto backfilling will be implemented as part of https://github.com/filecoin-project/lotus/issues/12116.

Stebalien replied:
Backfilling will only go back to the beginning of the local node's history.

aarshkshah1992 replied:
Fixed this by checking that we have processed events for the `minHeight` as well.

Stebalien commented:
FYI, https://github.com/filecoin-project/lotus/issues/12111

aarshkshah1992 commented on chain/events/filter/index.go:
```diff
@@ -98,6 +109,16 @@ const (
+	)`
+
+	createIndexEventsSeenHeight       = `CREATE INDEX IF NOT EXISTS events_seen_height ON events_seen (height);`
+	createIndexEventsSeenTipsetKeyCid = `CREATE INDEX IF NOT EXISTS events_seen_tipset_key_cid ON events_seen (tipset_key_cid);`
```
I don't really see a use case for it right now. We can add it when there is a need to.

Stebalien replied:
Probably fine? Assuming everything works and SQLITE doesn't squawk at you.

aarshkshah1992 commented on chain/events/filter/index.go:
```diff
@@ -75,20 +80,26 @@ var ddls = []string{
 	restoreEvent         = `UPDATE event SET reverted=false WHERE height=? AND tipset_key=? AND tipset_key_cid=? AND emitter_addr=? AND event_index=? AND message_cid=? AND message_index=?`
+	revertEventSeen      = `UPDATE events_seen SET reverted=true WHERE height=? AND tipset_key_cid=?`
+	restoreEventSeen     = `UPDATE events_seen SET reverted=false WHERE height=? AND tipset_key_cid=?`
+	insertEventsSeen     = `INSERT OR IGNORE INTO events_seen(height, tipset_key_cid, reverted) VALUES(?, ?, ?)`
```
Done.

aarshkshah1992 commented on chain/events/filter/index.go:
```diff
@@ -645,17 +855,66 @@ func (ei *EventIndex) CollectEvents(ctx context.Context, te *TipSetEvents, rever
 				if rowsAffected != 1 {
 					log.Warnf("restored %d events but expected only one to exist", rowsAffected)
+				} else {
+					isTipSetRevertRestored = true
```
Fixed this.

aarshkshah1992 commented on chain/events/filter/index.go:
```diff
@@ -401,9 +465,64 @@ func (ei *EventIndex) migrateToVersion5(ctx context.Context) error {
+    WHERE NOT EXISTS (
+        SELECT 1 FROM events_seen es
+        WHERE es.height = e.height AND es.tipset_key_cid = e.tipset_key_cid	
+    )
```
Fixed this.

aarshkshah1992 commented:
@Stebalien I have addressed all your review comments. 

Except for https://github.com/filecoin-project/lotus/pull/12080#discussion_r1645152589 which I would like to address as part of the refactor in https://github.com/filecoin-project/lotus/issues/12116 because that really fixes this in a clean way rather than patching the current broken dual write/dual read mechanism we have for (`ChainNotify` writes to filter buffer and index DB -> event APIs read from both filter buffer and index DB).

aarshkshah1992 commented:
@rvagg @Stebalien I'd like to merge this. Based on an offline conversation with @Stebalien -> he agrees this is good to go in it's current form as it does fix the missing event issue with`GetLogs`.

We will prioritise https://github.com/filecoin-project/lotus/issues/12116 next.

rvagg commented on chain/events/filter/index.go:
```diff
@@ -75,20 +80,27 @@ var ddls = []string{
 	restoreEvent         = `UPDATE event SET reverted=false WHERE height=? AND tipset_key=? AND tipset_key_cid=? AND emitter_addr=? AND event_index=? AND message_cid=? AND message_index=?`
+	revertEventSeen      = `UPDATE events_seen SET reverted=true WHERE height=? AND tipset_key_cid=?`
+	restoreEventSeen     = `UPDATE events_seen SET reverted=false WHERE height=? AND tipset_key_cid=?`
+	upsertEventsSeen     = `INSERT INTO events_seen(height, tipset_key_cid, reverted) VALUES(?, ?, ?) ON CONFLICT(height, tipset_key_cid) DO UPDATE SET reverted=false`
```
`reverted` is always `false` in this call, save a substitution by just setting it eh?

```suggestion
	upsertEventsSeen     = `INSERT INTO events_seen(height, tipset_key_cid, reverted) VALUES(?, ?, false) ON CONFLICT(height, tipset_key_cid) DO UPDATE SET reverted=false`
```

rvagg commented on node/impl/full/eth.go:
```diff
@@ -1258,10 +1263,60 @@ func (e *EthEventHandler) EthGetLogs(ctx context.Context, filterSpec *ethtypes.E
+			maxHeight = e.Chain.GetHeaviestTipSet().Height()
+		}
+		if maxHeight > e.Chain.GetHeaviestTipSet().Height() {
+			return nil, xerrors.Errorf("max height requested is greater than the heaviest tipset")
```
```suggestion
			return nil, xerrors.Errorf("maxHeight requested is greater than the heaviest tipset")
```

rvagg commented on node/impl/full/eth.go:
```diff
@@ -1258,10 +1263,60 @@ func (e *EthEventHandler) EthGetLogs(ctx context.Context, filterSpec *ethtypes.E
+		// should also have the minHeight in the filter indexed
+		b, err := e.EventFilterManager.EventIndex.IsHeightProcessed(ctx, uint64(pf.minHeight))
+		if err != nil {
+			return nil, xerrors.Errorf("failed to check if event index has events for the minheight: %w", err)
```
```suggestion
			return nil, xerrors.Errorf("failed to check if event index has events for the minHeight: %w", err)
```

rvagg commented on chain/events/filter/index.go:
```diff
@@ -651,11 +862,40 @@ func (ei *EventIndex) CollectEvents(ctx context.Context, te *TipSetEvents, rever
+	_, err = tx.Stmt(ei.stmtUpsertEventsSeen).Exec(
+		te.msgTs.Height(),
+		tsKeyCid.Bytes(),
+		false,
```
```suggestion
```

rvagg commented on node/impl/full/eth.go:
```diff
@@ -1270,6 +1325,53 @@ func (e *EthEventHandler) EthGetLogs(ctx context.Context, filterSpec *ethtypes.E
+	defer cancel()
+
+	// if the height we're interested in has already been indexed -> there's nothing to do here
+	b, err := ei.IsHeightProcessed(ctx, uint64(height))
+	if err != nil {
+		return xerrors.Errorf("failed to check if event index has events for given height: %w", err)
+	}
+	if b {
+		return nil
+	}
```
could compact this and the duplicate block below a little

```suggestion
	if b, err := ei.IsHeightProcessed(ctx, uint64(height)); err != nil {
		return xerrors.Errorf("failed to check if event index has events for given height: %w", err)
	} else if b {
		return nil
	}
```

(up to you tho)

aarshkshah1992 replied:
Reduced it below at multiple places . Thanks.

rvagg commented on node/impl/full/eth.go:
```diff
@@ -1270,6 +1325,53 @@ func (e *EthEventHandler) EthGetLogs(ctx context.Context, filterSpec *ethtypes.E
+	for {
+		select {
+		case <-subCh:
+			b, err = ei.IsHeightProcessed(ctx, uint64(height))
```
why don't we just return the height on the object coming from the channel and check that rather than needing to do a round-trip through the db at each epoch that's processed? we could have thousands of these queued up waiting for a far-future epoch, each one doing a SELECT on each epoch while they wait

rvagg replied:
oh, nevermind, `maxHeight > e.Chain.GetHeaviestTipSet().Height()` takes care of that concern
